Drop some dead code according to the clang static analyzer

cdumez opened the pull request:
#### e311e6b7fce5059373b719341561a759f6bae7d1
<pre>
Drop some dead code according to the clang static analyzer
<a href="https://bugs.webkit.org/show_bug.cgi?id=261658">https://bugs.webkit.org/show_bug.cgi?id=261658</a>

Reviewed by Darin Adler.

* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm:
(WebCore::ShapeDetection::configureRequestToUseCPUOrGPU):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::createSum):
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::markContextChangedAndNotifyCanvasObservers):
* Source/WebCore/inspector/InspectorFrontendClientLocal.cpp:
(WebCore::InspectorFrontendClientLocal::setAttachedWindow):
* Source/WebCore/layout/Verification.cpp:
(WebCore::Layout::outputMismatchingComplexLineInformationIfNeeded):
(WebCore::Layout::verifyAndOutputSubtree):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::isScrollable):
* Source/WebCore/platform/audio/ReverbConvolver.cpp:
(WebCore::ReverbConvolver::backgroundThreadEntry):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::fetchInternal):
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
* Source/WebCore/rendering/AutoTableLayout.cpp:
(WebCore::AutoTableLayout::calcEffectiveLogicalWidth):
* Source/WebCore/rendering/LegacyLineLayout.cpp:
(WebCore::constructBidiRunsForSegment):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::width const):
* Source/WebCore/rendering/mathml/MathOperator.cpp:
(WebCore::MathOperator::calculateStretchyData):
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::endPointsFromAngleForPrefixedVariants):
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::computedUnderlineOffset):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::parserMetaData):
* Source/WebCore/xml/XSLTUnicodeSort.cpp:
(WebCore::xsltUnicodeSortFunction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: <a href="https://commits.webkit.org/268092@main">https://commits.webkit.org/268092@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c05f756ea1cc0ef559442035986e8b49fdf2fe22

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18652 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18992 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19595 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20514 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17459 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19133 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19308 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19033 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16239 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21390 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16999 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23448 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17281 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21341 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17773 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15082 "4 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16826 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4435 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21193 "Built successfully") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/2287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1770,7 +1770,8 @@ float RenderText::width(unsigned from, unsigned length, const FontCascade& fontC
                 if (!m_maxWidth)
                     width = maxLogicalWidth();
-                width = *m_maxWidth;
+                else
```
@alanbaradlay FYI, this is one place where I fixed the code rather than dropping the dead code.

darinadler replied:
Looks like this code would crash or abort with a null dereference before? Any ideas how to make a test case?

cdumez replied:
@alanbaradlay What do you think?

alanbaradlay replied:
not great code (especially the const_cast part) but calling maxLogicalWidth() does assign a value to m_maxWidth.
```
float RenderText::maxLogicalWidth() const
{
    if (preferredLogicalWidthsDirty() || !m_maxWidth)
        const_cast<RenderText*>(this)->computePreferredLogicalWidths(0, !preferredLogicalWidthsDirty());

    return *m_maxWidth;
}

```

cdumez replied:
Oh wow, Ok. So, maybe a a static analysis false positive.

I guess we should refactor this code to make this clearer and hopefully get rid of the static analysis warning.

alanbaradlay commented on Source/WebCore/rendering/LegacyLineLayout.cpp:
```diff
@@ -1174,7 +1174,7 @@ static inline void constructBidiRunsForSegment(InlineBidiResolver& topResolver,
     ASSERT(&topResolver.runs() == &bidiRuns);
     ASSERT(topResolver.position() != endOfRuns);
-    RenderObject* currentRoot = topResolver.position().root();
+    RenderObject* currentRoot;
```
maybe RenderObject* currentRoot = nullptr; ?
or
```
@@ -1174,14 +1174,12 @@ static inline void constructBidiRunsForSegment(InlineBidiResolver& topResolver,
     // of the resolver owning the runs.
     ASSERT(&topResolver.runs() == &bidiRuns);
     ASSERT(topResolver.position() != endOfRuns);
-    RenderObject* currentRoot = topResolver.position().root();
     topResolver.createBidiRunsForLine(endOfRuns, override, previousLineBrokeCleanly);
 
     while (!topResolver.isolatedRuns().isEmpty()) {
         // It does not matter which order we resolve the runs as long as we resolve them all.
         auto isolatedRun = WTFMove(topResolver.isolatedRuns().last());
         topResolver.isolatedRuns().removeLast();
-        currentRoot = &isolatedRun.root;
 
         RenderObject& startObject = isolatedRun.object;
 
@@ -1190,7 +1188,7 @@ static inline void constructBidiRunsForSegment(InlineBidiResolver& topResolver,
         // tree to see which parent inline is the isolate. We could change enterIsolate
         // to take a RenderObject and do this logic there, but that would be a layering
         // violation for BidiResolver (which knows nothing about RenderObject).
-        RenderInline* isolatedInline = downcast<RenderInline>(highestContainingIsolateWithinRoot(startObject, currentRoot));
+        RenderInline* isolatedInline = downcast<RenderInline>(highestContainingIsolateWithinRoot(startObject, &isolatedRun.root));
         ASSERT(isolatedInline);
 
         InlineBidiResolver isolatedResolver;

```

cdumez replied:
Fixed

darinadler commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -227,7 +227,7 @@ ALWAYS_INLINE bool matchesLangPseudoClass(const Element& element, const FixedVec
         unsigned rangeLength = rangeStringView.length();
         unsigned rangeSubtagsStartIndex = 0;
-        unsigned rangeSubtagsEndIndex = rangeLength;
+        unsigned rangeSubtagsEndIndex;
```
To make things clearer, we could move this definition inside the loop.

darinadler commented on Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
```diff
@@ -218,7 +218,7 @@ - (void)interruption:(NSNotification *)notification
     }
 
-    NSString *modeString = AVAudioSessionModeDefault;
+    NSString *modeString;
```
Unless this is compiled with ARC, this makes the code have an uninitialized string pointer if `newMode` is a bogus enum value. Are we OK with that?

cdumez replied:
Will tweak to avoid this.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -183,7 +183,9 @@ void Font::platformInit()
 
     // Compute line spacing before the line metrics hacks are applied.
+#if !PLATFORM(IOS_FAMILY)
     float lineSpacing = lroundf(ascent) + lroundf(descent) + lroundf(lineGap);
```
The use of `lroundf` here is a bug dating back to when `CGFloat` was 32-bit, causing these numbers to all be converted to `float` before being rounded. It would be better to use `std::lround` here, and fixing it seems low risk.

darinadler commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -201,7 +203,7 @@ void Font::platformInit()
     lineGap = ceilf(lineGap);
-    lineSpacing = ceil(ascent) + adjustment + ceil(descent) + lineGap;
+    float lineSpacing = ceil(ascent) + adjustment + ceil(descent) + lineGap;
     ascent = ceilf(ascent + adjustment);
```
Similar error here with `ceilf`.

darinadler replied:
The mix of `float` and `double` here seems sloppy, is different between 32-bit and 64-bit because of `CGFloat`, slower to convert back and forth, but likely harmless. Since the data in WebCore proper is kept as `float` it’s not like we get a significant benefit from higher precision in the code here.

cdumez committed:
Drop some dead code according to the clang static analyzer
https://bugs.webkit.org/show_bug.cgi?id=261658

Reviewed by Darin Adler.

* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/VisionUtilities.mm:
(WebCore::ShapeDetection::configureRequestToUseCPUOrGPU):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass):
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::createSum):
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::markContextChangedAndNotifyCanvasObservers):
* Source/WebCore/inspector/InspectorFrontendClientLocal.cpp:
(WebCore::InspectorFrontendClientLocal::setAttachedWindow):
* Source/WebCore/layout/Verification.cpp:
(WebCore::Layout::outputMismatchingComplexLineInformationIfNeeded):
(WebCore::Layout::verifyAndOutputSubtree):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::isScrollable):
* Source/WebCore/platform/audio/ReverbConvolver.cpp:
(WebCore::ReverbConvolver::backgroundThreadEntry):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::fetchInternal):
* Source/WebCore/platform/audio/ios/AudioSessionIOS.mm:
(WebCore::AudioSessionIOS::setCategory):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
* Source/WebCore/rendering/AutoTableLayout.cpp:
(WebCore::AutoTableLayout::calcEffectiveLogicalWidth):
* Source/WebCore/rendering/LegacyLineLayout.cpp:
(WebCore::constructBidiRunsForSegment):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::width const):
* Source/WebCore/rendering/mathml/MathOperator.cpp:
(WebCore::MathOperator::calculateStretchyData):
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::endPointsFromAngleForPrefixedVariants):
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::computedUnderlineOffset):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::parserMetaData):
* Source/WebCore/xml/XSLTUnicodeSort.cpp:
(WebCore::xsltUnicodeSortFunction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didAttachToRunningProcess):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: https://commits.webkit.org/268092@main
